src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,472 @@
+#pragma once
+
+#include "mfem.hpp"
+
+#include "serac/physics/utilities/variational_form/tensor.hpp"
+#include "serac/physics/utilities/variational_form/quadrature.hpp"
+#include "serac/physics/utilities/variational_form/finite_element.hpp"
+#include "serac/physics/utilities/variational_form/tuple_arithmetic.hpp"
+#include "serac/physics/utilities/variational_form/integral.hpp"
+
+namespace serac {
+
+/// @cond
+template <typename T>
+class WeakForm;
+/// @endcond
+
+/**
+ * @brief Intended to be like @p std::function for finite element kernels
+ *
+ * That is: you tell it the inputs (trial spaces) for a kernel, and the outputs (test space) like @p std::function
+ * For example, this code represents a function that takes an integer argument and returns a double:
+ * @code{.cpp}
+ * std::function< double(double, int) > my_func;
+ * @endcode
+ * And this represents a function that takes values from an Hcurl field and returns a
+ * residual vector associated with an H1 field:
+ * @code{.cpp}
+ * WeakForm< H1(Hcurl) > my_residual;
+ * @endcode
+ *
+ * @tparam test The space of test functions to use
+ * @tparam trial The space of trial functions to use
+ *
+ * To use this class, you use the methods @p WeakForm::Add****Integral(integrand,domain_of_integration)
+ * where @p integrand is a q-function lambda or functor and @p domain_of_integration is an @p mfem::mesh
+ *
+ * @see https://libceed.readthedocs.io/en/latest/libCEEDapi/#theoretical-framework for additional
+ * information on the idea behind a quadrature function and its inputs/outputs
+ *
+ * @code{.cpp}
+ * // for domains made up of quadrilaterals embedded in R^2
+ * my_residual.AddAreaIntegral(integrand, domain_of_integration);
+ * // alternatively...
+ * my_residual.AddDomainIntegral(Dimension<2>{}, integrand, domain_of_integration);
+ *
+ * // for domains made up of quadrilaterals embedded in R^3
+ * my_residual.AddSurfaceIntegral(integrand, domain_of_integration);
+ *
+ * // for domains made up of hexahedra embedded in R^3
+ * my_residual.AddVolumeIntegral(integrand, domain_of_integration);
+ * // alternatively...
+ * my_residual.AddDomainIntegral(Dimension<3>{}, integrand, domain_of_integration);
+ * @endcode
+ */
+template <typename test, typename trial>
+class WeakForm<test(trial)> : public mfem::Operator {
+public:
+  /**
+   * @brief Constructs using @p mfem::ParFiniteElementSpace objects corresponding to the test/trial spaces
+   * @param[in] test_fes The test space
+   * @param[in] trial_fes The trial space
+   */
+  WeakForm(mfem::ParFiniteElementSpace* test_fes, mfem::ParFiniteElementSpace* trial_fes)
```

Comment:
Unfortunately, I don't see a way to get away with only one or the other (without making serious changes to mfem). Fundamentally, mfem makes the choice of H1, Hcurl, Hdiv, etc is a runtime one, while I require us to make that choice at compile time. The reason for wanting to know this information at compile time has to do with the fact that a q-function doesn't make sense if you try and apply it to the wrong test/trial space. For example, let's look at a q-function for isotropic linear elasticity:

```cpp
[](auto displacement) {
  static constexpr auto I = Identity<dim>();
  auto [u, du_dx] = displacement;
  auto body_force = 2.0 * u;
  auto strain = 0.5 * (du_dx + transpose(du_dx));
  auto stress = 3.0 * tr(strain) * I + 4.0 * strain;
  return std::tuple{body_force, stress};
};
```

It doesn't make sense to apply these operations to, say, a temperature field. The temperature gradient is a vector, rather than a matrix, so `transpose` and `tr` aren't defined, and the shapes of all the inputs and outputs are wrong.

We need the information from `mfem::ParFiniteElementSpace` as soon as possible to initialize the prolongation / restriction operators that are applied before and after our templated finite element pipeline. So, I require the finite element spaces in the constructor.

Reply:
Yeah, that makes sense. Thanks for the explanation!